File: sdk/python/ferrite_ai/langchain.py
"""LangChain integration — use Ferrite as a semantic cache for LLM calls.

Usage::

    from langchain.globals import set_llm_cache
    from ferrite_ai.langchain import FerriteSemanticCache

    set_llm_cache(FerriteSemanticCache())
"""

from __future__ import annotations

import hashlib
import json
from typing import Any, Dict, List, Optional, Sequence, Union

from ferrite_ai.client import FerriteClient

try:
    from langchain_core.caches import BaseCache, RETURN_VAL_TYPE
    from langchain_core.outputs import Generation

    _HAS_LANGCHAIN = True
except ImportError:  # pragma: no cover
    _HAS_LANGCHAIN = False

    class BaseCache:  # type: ignore[no-redef]
        """Stub so the module can be imported without LangChain."""

        def lookup(self, prompt: str, llm_string: str) -> Any: ...
        def update(self, prompt: str, llm_string: str, return_val: Any) -> None: ...
        def clear(self, **kwargs: Any) -> None: ...

    RETURN_VAL_TYPE = Any  # type: ignore[assignment,misc]


class FerriteSemanticCache(BaseCache):
    """LangChain-compatible semantic cache backed by Ferrite.

    Parameters
    ----------
    host : str
        Ferrite host (default ``"127.0.0.1"``).
    port : int
        Ferrite port (default ``6380``).
    similarity_threshold : float
        Minimum similarity to consider a cache hit (default ``0.85``).
    ttl : int | None
        Time-to-live in seconds for cached entries.
    namespace : str
        Key prefix for cache isolation (default ``"lc_cache"``).
    client : FerriteClient | None
        Pre-configured :class:`FerriteClient`.  When provided, *host* / *port*
        / *namespace* are ignored.
    """

    def __init__(
        self,
        host: str = "127.0.0.1",
        port: int = 6380,
        similarity_threshold: float = 0.85,
        ttl: Optional[int] = None,
        namespace: str = "lc_cache",
        client: Optional[FerriteClient] = None,
    ) -> None:
        if not _HAS_LANGCHAIN:
            raise ImportError(
                "LangChain is required for this integration. "
                "Install it with: pip install langchain-core"
            )
        self._client = client or FerriteClient(
            host=host, port=port, namespace=namespace,
        )
        self.similarity_threshold = similarity_threshold
        self.ttl = ttl

    def lookup(self, prompt: str, llm_string: str) -> Optional[RETURN_VAL_TYPE]:
        """Look up a cached LLM response for *prompt*.

        Returns a list of :class:`Generation` objects on a hit, or ``None``.
        """
        cache_key = self._make_key(prompt, llm_string)
        result = self._client.semantic_get(cache_key, threshold=self.similarity_threshold)
        if result is None:
            return None
        try:
            generations_data = json.loads(result.response)
            return [Generation(**g) for g in generations_data]
        except (json.JSONDecodeError, TypeError):
            return [Generation(text=result.response)]

    def update(self, prompt: str, llm_string: str, return_val: RETURN_VAL_TYPE) -> None:
        """Store an LLM response in the semantic cache."""
        cache_key = self._make_key(prompt, llm_string)
        generations_data = [
            {"text": g.text, "generation_info": g.generation_info}
            for g in return_val
        ]
        self._client.semantic_set(
            query=cache_key,
            response=json.dumps(generations_data),
            metadata={"llm": llm_string[:128]},
            ttl=self.ttl,
        )

    def clear(self, **kwargs: Any) -> None:
        """Clear all cached entries (best-effort)."""
        # Delegate to Ferrite; a namespace-scoped flush is not part of the
        # SEMANTIC command set yet, so we issue a pattern-based delete.
        try:
            self._client.redis.execute_command(
                "SEMANTIC.FLUSH", self._client.namespace,
            )
        except Exception:
            pass

    # -- helpers -------------------------------------------------------------

    @staticmethod
    def _make_key(prompt: str, llm_string: str) -> str:
        """Combine prompt and LLM identifier into a single cache key."""
        tag = hashlib.sha256(llm_string.encode()).hexdigest()[:8]
        return f"{tag}:{prompt}"

File: sdk/typescript/src/langchain.ts
/**
 * LangChain.js integration — use Ferrite as a semantic cache for LLM calls.
 *
 * @example
 * ```ts
 * import { FerriteSemanticCache } from "@ferrite/ai";
 * import { ChatOpenAI } from "@langchain/openai";
 *
 * const cache = new FerriteSemanticCache({ similarityThreshold: 0.85 });
 * const llm = new ChatOpenAI({ model: "gpt-4o-mini", cache });
 * ```
 */

import { createHash } from "crypto";
import { FerriteClient, type FerriteClientOptions } from "./client";

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

export interface FerriteSemanticCacheOptions extends FerriteClientOptions {
  similarityThreshold?: number;
  ttl?: number;
  client?: FerriteClient;
}

/** Minimal Generation shape expected by LangChain caches. */
interface Generation {
  text: string;
  generationInfo?: Record<string, unknown>;
}

// ---------------------------------------------------------------------------
// Cache
// ---------------------------------------------------------------------------

/**
 * LangChain.js-compatible semantic cache backed by Ferrite.
 *
 * Implements the `lookup` / `update` / `clear` interface expected by
 * `@langchain/core` BaseCache.
 */
export class FerriteSemanticCache {
  private readonly client: FerriteClient;
  private readonly similarityThreshold: number;
  private readonly ttl?: number;

  constructor(options: FerriteSemanticCacheOptions = {}) {
    const { similarityThreshold = 0.85, ttl, client, ...clientOpts } = options;

    this.client =
      client ??
      new FerriteClient({
        ...clientOpts,
        namespace: clientOpts.namespace ?? "lc_cache",
      });
    this.similarityThreshold = similarityThreshold;
    this.ttl = ttl;
  }

  /**
   * Look up a cached LLM response for the given prompt.
   *
   * @returns An array of Generation objects on a hit, or `null`.
   */
  async lookup(
    prompt: string,
    llmString: string,
  ): Promise<Generation[] | null> {
    const cacheKey = this.makeKey(prompt, llmString);
    const result = await this.client.semanticGet(
      cacheKey,
      this.similarityThreshold,
    );
    if (result === null) {
      return null;
    }

    try {
      const data = JSON.parse(result.response) as Array<{
        text: string;
        generationInfo?: Record<string, unknown>;
      }>;
      return data.map((g) => ({
        text: g.text,
        generationInfo: g.generationInfo,
      }));
    } catch {
      return [{ text: result.response }];
    }
  }

  /**
   * Store an LLM response in the semantic cache.
   */
  async update(
    prompt: string,
    llmString: string,
    returnVal: Generation[],
  ): Promise<void> {
    const cacheKey = this.makeKey(prompt, llmString);
    const data = returnVal.map((g) => ({
      text: g.text,
      generationInfo: g.generationInfo,
    }));
    await this.client.semanticSet(
      cacheKey,
      JSON.stringify(data),
      { llm: llmString.slice(0, 128) },
      this.ttl,
    );
  }

  /**
   * Clear all cached entries (best-effort).
   */
  async clear(): Promise<void> {
    try {
      await this.client.ioRedis.call("SEMANTIC.FLUSH", "lc_cache");
    } catch {
      // flush not supported yet
    }
  }

  private makeKey(prompt: string, llmString: string): string {
    const tag = createHash("sha256")
      .update(llmString)
      .digest("hex")
      .slice(0, 8);
    return `${tag}:${prompt}`;
  }
}
